### ISSUE TITLE
send key event action when imitating button presses

### ISSUE THREAD

**Issue by yyy33** [sds100/KeyMapper#563]
Hello sds100, I use `rd client` to connect to the remote desktop on my Android phone, and then use the Bluetooth keyboard to connect to my Android phone. But I found that there are some problems when using the keyboard, that is, when I press certain shortcut keys on the keyboard, these shortcut keys will be triggered by the Android system instead of being triggered by the remote computer, such as `alt-tab`, `meta`.

In order to solve this problem, I checked a lot of information and finally found the software you developed. I set up a pair of mappings in the software: `short press meta key`-----`meta key`. At this time, my remote computer can receive the meta key normally, but cannot use the key combination, such as opening the windows file manager `meta-e`. Can you teach me how to configure it?

**Comment by sds100** [comment]
Hi! I'm not 100% what the solution is but there is one thing I can think of trying. Can you send a screenshot of your keymaps as they are now?
1. Install the apk ("artifact") from this CI build https://github.com/sds100/KeyMapper/actions/runs/498957297. 
2. Then create a "Key Event" action and select your keyboard as the device for it. The screenshot below will show you what the screen to create this action will look like.

![Screenshot_20210121-131803](https://user-images.githubusercontent.com/16245954/105356341-2e782500-5beb-11eb-84ac-4c268b42daf8.png)

**Comment by yyy33** [comment]
> Hi! I'm not 100% what the solution is but there is one thing I can think of trying. Can you send a screenshot of your keymaps as they are now?
> 
> 1. Install the apk ("artifact") from this CI build https://github.com/sds100/KeyMapper/actions/runs/498957297.
> 2. Then create a "Key Event" action and select your keyboard as the device for it. The screenshot below will show you what the screen to create this action will look like.
> 
> ![Screenshot_20210121-131803](https://user-images.githubusercontent.com/16245954/105356341-2e782500-5beb-11eb-84ac-4c268b42daf8.png)

Hello, I downloaded and installed the installation package file you provided. The following screenshot is my configuration, but I still can't use key combination. I don't know what the problem is.
![Screenshot_20210122-192820](https://user-images.githubusercontent.com/51548147/105486314-4209b580-5ce9-11eb-8ecf-4106d85592e6.jpg)
![Screenshot_20210122-192829](https://user-images.githubusercontent.com/51548147/105486361-52219500-5ce9-11eb-9071-e6950fccbd11.jpg)
![Screenshot_20210122-192839](https://user-images.githubusercontent.com/51548147/105486375-56e64900-5ce9-11eb-810f-6aed19fcf75b.jpg)
![Screenshot_20210122-192847](https://user-images.githubusercontent.com/51548147/105486388-5b126680-5ce9-11eb-9190-300422aea6de.jpg)
![Screenshot_20210122-192856](https://user-images.githubusercontent.com/51548147/105486400-5f3e8400-5ce9-11eb-97e1-42f4d75837d7.jpg)
![Uploading Screenshot_20210122-192907.jpg…]()
![Uploading Screenshot_20210122-193303.jpg…]()

**Comment by sds100** [comment]
What if you turn on the "hold down" option by pressing the 3 dots on the action.

**Comment by yyy33** [comment]
> What if you turn on the "hold down" option by pressing the 3 dots on the action.

Is it set up like this? Still the same as before.

![Screenshot_20210123-084347](https://user-images.githubusercontent.com/51548147/105562925-2f2dca00-5d57-11eb-86de-9e9cade6875a.jpg)

**Comment by sds100** [comment]
You didn't have the hold down option turned on before

**Comment by yyy33** [comment]
> You didn't have the hold down option turned on before.
> And can you choose a device for the Key Event action like in my screenshot.

Yes, I can see my Bluetooth keyboard in the device list, and I chose it.

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5859506/keymapper-2.3.0-alpha-debug.zip)
Uninstall the old debug build and install this one.
This apk will intercept **all** key events and then send them through the Key Mapper Debug Basic Input Method.
Turn on the accessibility service and you must choose the Key Mapper Debug Basic Input Method. You know it is working if your volume buttons aren't working. You don't need to create any keymaps just press the shortcut into your rd client. I want to see if passing along the system's key events will work.

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5859506/keymapper-2.3.0-alpha-debug.zip)
> Uninstall the old debug build and install this one.
> This apk will intercept **all** key events and then send them through the Key Mapper Debug Basic Input Method.
> Turn on the accessibility service and you must choose the Key Mapper Debug Basic Input Method. You know it is working if your volume buttons aren't working. You don't need to create any keymaps just press the shortcut into your rd client. I want to see if passing along the system's key events will work.

Ha ha, it works. The remote computer can normally receive all key sequences on my Bluetooth keyboard, including combination keys.

**Comment by sds100** [comment]
Cool. That's a good start. Can you install this app and press the key shortcuts with the accessibility service of the last apk turned on and then with it off. Then send me a screenshot of the full output.
[keyevent-viewer-deviceid-source.zip](https://github.com/sds100/KeyMapper/files/5860265/keyevent-viewer-deviceid-source.zip)

**Comment by yyy33** [comment]
> Cool. That's a good start. Can you install this app and press the key shortcuts with the accessibility service of the last apk turned on and then with it off. Then send me a screenshot of the full output.
> [keyevent-viewer-deviceid-source.zip](https://github.com/sds100/KeyMapper/files/5860265/keyevent-viewer-deviceid-source.zip)

Of course, when the accessibility service is turned on, the software does not have any output. When the accessibility service is turned off, if I press meta-e quickly, the following output will be generated. However, if the speed is not very fast, there will be no output
````
Press buttons to see the keycodes:
KeyUp       : action=1 code=33 repeat=0 meta=0 scanCode=18 flags=8 label='69' chars='null' number='0' deviceId='9' source='769'
````

**Comment by sds100** [comment]
Did you choose the Basic Input Method when you turned the accessibility service on? it should work.

**Comment by yyy33** [comment]
> Did you choose the Basic Input Method when you turned the accessibility service on? it should work.

No,i don't.Let me try again 
````
Press buttons to see the keycodes:
KeyDown     : action=0 code=117 repeat=0 meta=196608 scanCode=125 flags=0 label='0' chars='null' number='0' deviceId='10' source='769'
KeyDown     : action=0 code=33 repeat=0 meta=196608 scanCode=18 flags=0 label='69' chars='null' number='0' deviceId='10' source='769'
KeyUp       : action=1 code=33 repeat=0 meta=196608 scanCode=18 flags=0 label='69' chars='null' number='0' deviceId='10' source='769'
KeyUp       : action=1 code=117 repeat=0 meta=0 scanCode=125 flags=0 label='0' chars='null' number='0' deviceId='10' source='769'
-------------------------
-------------------------
-------------------------
KeyUp       : action=1 code=33 repeat=0 meta=0 scanCode=18 flags=8 label='69' chars='null' number='0' deviceId='10' source='769'

````

**Comment by sds100** [comment]
How come there is only a KeyUp if you press the shortcut without the accessibility service turned on? There should be 4 events - down and up for both Meta and E

**Comment by yyy33** [comment]
> How come there is only a KeyUp if you press the shortcut without the accessibility service turned on? There should be 4 events - down and up for both Meta and E

I tried many times and there was only one. And I'm still pressing the shortcut key quickly. If I press a little bit slowly, nothing will be output. Is it related to voice assistant? When I press the meta key alone, the voice assistant is turned on and there is no output.

**Comment by sds100** [comment]
ohhhh, of course! sorry

**Comment by yyy33** [comment]
> How come there is only a KeyUp if you press the shortcut without the accessibility service turned on? There should be 4 events - down and up for both Meta and E

My device has root permissions. I have given root permissions to the software that tests the buttons.

**Comment by sds100** [comment]
yeah, its fine. i forgot that the keyevents were being consumed by the system 😄

**Comment by sds100** [comment]
I'm working on a fix now...

**Comment by yyy33** [comment]
> I'm working on a fix now...

Hard work, come on.🌹

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5860448/keymapper-2.3.0-alpha-debug.zip)
try this apk. do what you were first doing by remapping the meta key to the meta key. you **_must_** have the "hold down" option turned on for the meta key code action. you don't need to set the device or anything.

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5860448/keymapper-2.3.0-alpha-debug.zip)
> try this apk. do what you were first doing by remapping the meta key to the meta key. you **_must_** have the "hold down" option turned on for the meta key code action. you don't need to set the device or anything.

Remote desktop can receive shortcut keys normally. But I found some small problems. The first one is that when I press the shortcut key meta-e at a very fast speed, the effect on the remote desktop is equivalent to pressing meta alone, then releasing, then pressing E alone, and then releasing. The second problem is that if I use the Bluetooth keyboard to directly connect to my remote computer, at this time I keep pressing and holding the meta key. Only when I release it will the windows application menu interface be opened. But when I use the Bluetooth keyboard to connect to the Android phone and use the Android phone to connect to my computer remotely, if I keep holding down the meta key, before I release it, the windows application menu interface will be opened in two or three seconds.

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5861482/keymapper-2.3.0-alpha-debug.zip)
this should have fixed the issues

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5861482/keymapper-2.3.0-alpha-debug.zip)
> this should have fixed the issues

When I install, I am prompted that the installation package is invalid. No operating system is incompatible. My operating system is 8.0😂

**Comment by sds100** [comment]
ohh oops. sent wrong one i think...

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5861530/keymapper-2.3.0-alpha-debug.zip)
does this work?

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5861530/keymapper-2.3.0-alpha-debug.zip)
> does this work?

This installation package can't use shortcut keys, just like the release version.

**Comment by sds100** [comment]
You're using the basic input method?

**Comment by yyy33** [comment]
> You're using the basic input method?

yes ，i do。

**Comment by sds100** [comment]
[Uploading keymapper-2.3.0-alpha-debug_1.zip…]()
try this apk

**Comment by yyy33** [comment]
> try this apk
> [keymapper-2.3.0-alpha-debug_1.zip](https://github.com/sds100/KeyMapper/files/5862557/keymapper-2.3.0-alpha-debug_1.zip)

It's the same as the release version.

**Comment by sds100** [comment]
Really?! 😂 It passes the same tests as the first one I sent which worked. It has the same output in Key Event display as you sent earlier.

**Comment by yyy33** [comment]
Yes😂, I tried twice and couldn't use the shortcut keys. When I got home, I tried the previous version that could use the shortcut keys.

**Comment by yyy33** [comment]
> Really?! 😂 It passes the same tests as the first one I sent which worked. It has the same output in Key Event display as you sent earlier.

Yes😂, I tried twice and couldn't use the shortcut keys. When I got home, I tried the previous version that could use the shortcut keys.

**Comment by yyy33** [comment]
> Really?! 😂 It passes the same tests as the first one I sent which worked. It has the same output in Key Event display as you sent earlier.

Has this installation package been modified by you? It's no longer able to use shortcuts.
![Uploading IMG_20210125_203315.jpg…]()

**Comment by sds100** [comment]
You have the "hold down" option turned on right?

**Comment by yyy33** [comment]
> You have the "hold down" option turned on right?

Yes, i have

**Comment by sds100** [comment]
![Screenshot_20210125-180838](https://user-images.githubusercontent.com/16245954/105747200-67403300-5f38-11eb-9d7b-fd50aa712096.png)
are you using this keymap and are you using the Basic Input Method?

**Comment by sds100** [comment]
idk what the problem is because it outputs the exact same as the Key Event Display output that you sent me

**Comment by yyy33** [comment]
> ![Screenshot_20210125-180838](https://user-images.githubusercontent.com/16245954/105747200-67403300-5f38-11eb-9d7b-fd50aa712096.png)
> are you using this keymap and are you using the Basic Input Method?

Yes, I have. Here is my configuration picture.
I found a problem that even though I granted root permission to the software, it still prompted me to fix something.

![Screenshot_20210126-073125](https://user-images.githubusercontent.com/51548147/105788472-8e8c1400-5fbb-11eb-9a23-f077fb121dd7.jpg)
![Screenshot_2021-01-26-07-31-00](https://user-images.githubusercontent.com/51548147/105788480-9350c800-5fbb-11eb-8277-0b3c9aeb69f7.jpg)
![Screenshot_20210126-072938](https://user-images.githubusercontent.com/51548147/105788488-96e44f00-5fbb-11eb-8355-5627953c5f28.jpg)
![Screenshot_20210126-072930](https://user-images.githubusercontent.com/51548147/105788494-99df3f80-5fbb-11eb-819e-3765d1de1972.jpg)
![Screenshot_20210126-072920](https://user-images.githubusercontent.com/51548147/105788502-9d72c680-5fbb-11eb-8da1-cdc63b604cfe.jpg)
![Screenshot_20210126-072820](https://user-images.githubusercontent.com/51548147/105788511-a1064d80-5fbb-11eb-9fd5-cb3d8e6eb217.jpg)
![Screenshot_20210126-070746](https://user-images.githubusercontent.com/51548147/105788528-a4013e00-5fbb-11eb-8349-a5596a62be0e.jpg)
![Screenshot_20210126-070730](https://user-images.githubusercontent.com/51548147/105788541-a794c500-5fbb-11eb-9d32-3c9c20aee271.jpg)

**Comment by yyy33** [comment]
> idk what the problem is because it outputs the exact same as the Key Event Display output that you sent me

Can you use shortcut keys on the remote desktop of your mobile phone? If your mobile phone is OK, I think it's something wrong with my mobile phone.

**Comment by sds100** [comment]
turn off repeat! that messes it up

**Comment by yyy33** [comment]
> turn off repeat! that messes it up. hopefully that is the issue 😀

I'll try it when I get home, but I didn't turn off "repeat" the last time I could use the shortcut key normally.😂

**Comment by yyy33** [comment]
> turn off repeat! that messes it up. hopefully that is the issue 😀

The bad news is that I still can't use the shortcut after turning off the `repeat` option.🤣

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5875010/keymapper-2.3.0-alpha-debug.zip)
I tested this on remote desktop and this works. I need to pass through the scancode for the Key Event. I don't know how you got one of the apks to work 😂

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5875010/keymapper-2.3.0-alpha-debug.zip)
> I tested this on remote desktop and this works. I need to pass through the scancode for the key event. I don't know how you got one of the apks to work 😂

Invalid installation package, no operating system incompatibility, I received this prompt when installing

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5877060/keymapper-2.3.0-alpha-debug.zip)
does this work?

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug.zip](https://github.com/sds100/KeyMapper/files/5877060/keymapper-2.3.0-alpha-debug.zip)
> does this work?

Or can only short press, can't use shortcut keys.😴

**Comment by sds100** [comment]
What! 😂 I swear it worked in the Microsoft Remote Desktop app for me. I could use Win+E to open File Explorer and Win+R to open the Run dialog.

**Comment by yyy33** [comment]
> What! 😂 I swear it worked in the Microsoft Remote Desktop app for me. I could use Win+E to open File Explorer and Win+R to open the Run dialog.

😭Is there something wrong with my mobile phone? I remember I succeeded once. But there's a little bit of a problem. That's the one.

**Comment by sds100** [comment]
Did you change anything? You were right, the other apks didn't work for me either. The one you said that worked didn't seem to either.

**Comment by yyy33** [comment]
This version can work normally, but there are some small problems I mentioned.

> Did you change anything? You were right, the other apks didn't work for me either. The one you said that worked didn't seem to either. The one which just passed through all Key Events did work.

This version can work normally, but there are some small problems I mentioned.
![Uploading IMG_20210127_211246.jpg…]()

**Comment by sds100** [comment]
Can you make sure the image is uploaded before posting the comment. I've done the same as well

**Comment by yyy33** [comment]
> Can you make sure the image is uploaded before posting the comment. I've done the same as well

Can you see the picture now? My mobile browser doesn't seem very smart
![Uploading IMG_20210127_211246.jpg…]()

**Comment by sds100** [comment]
Nope at first the link will say "Uploading [image name]...." and when it doesn't say "Uploading" anymore you need to send the message.

**Comment by sds100** [comment]
Closed?

**Comment by yyy33** [comment]
> Closed?

No, my browser is too laggy. It should be wrong.

**Comment by yyy33** [comment]
> Nope at first the link will say "Uploading [image name]...." and when it doesn't say "Uploading" anymore you need to send the message.
Now you should be able to see the pictures. I uploaded them on my computer instead.
![Photo_0127_1a](https://user-images.githubusercontent.com/51548147/105998072-3baf7b00-60e7-11eb-8308-38d1aeca09a9.jpg)

**Comment by sds100** [comment]
So you can't use the shortcuts _at all_ in the latest apk that I sent?

**Comment by yyy33** [comment]
> So you can you use the shortcuts _at all_ in the latest apk that I sent?

No, only the installation package in the picture I sent can use shortcut keys.

**Comment by sds100** [comment]
[keymapper-2.3.0-alpha-debug_1.zip](https://github.com/sds100/KeyMapper/files/5880584/keymapper-2.3.0-alpha-debug_1.zip)
I sent you the wrong apk. 😂 Sorry that this is taking so long

**Comment by yyy33** [comment]
> [keymapper-2.3.0-alpha-debug_1.zip](https://github.com/sds100/KeyMapper/files/5880584/keymapper-2.3.0-alpha-debug_1.zip)
> I sent you the wrong apk. 😂 Sorry that this is taking so long

😂😂😂
The shortcut key is normal, but I found that there are two shortcut keys with a little problem. `meta -enter` and `meta-backspace` are used as `home key` and `back key` in my mobile phone system. When I turn on mapping, other shortcut keys of remote desktop are normal, but as soon as I press the above two shortcut keys, Android system will respond to these two shortcut keys, and will respond repeatedly. As a result, my mobile phone has been pressing the `home key` or `back key`. Just now my mobile phone couldn't do anything until I forced it to turn off.😥

**Comment by sds100** [comment]
Do you need to use Meta-Enter as a shortcut?

**Comment by sds100** [comment]
![Screenshot_20210127-173156589](https://user-images.githubusercontent.com/16245954/106029998-989a3980-60c5-11eb-8dac-5b1228121520.jpg)
Make these shortcuts by pressing meta-enter to create a "home" trigger and meta-backspace to create a "back" trigger. I could open Narrator on Windows with the Ctrl+meta+enter shortcut after doing this.

**Comment by yyy33** [comment]
> Do you need to use Meta-Enter as a shortcut?

Well, when I use VNC to connect to Linux, I often need to use `meta-enter` to open the shell

**Comment by yyy33** [comment]
> ![Screenshot_20210127-173156589](https://user-images.githubusercontent.com/16245954/106029998-989a3980-60c5-11eb-8dac-5b1228121520.jpg)
> Make these shortcuts by pressing meta-enter to create a "home" trigger and meta-backspace to create a "back" trigger. I could open Narrator on Windows with the Ctrl+meta+enter shortcut after doing this.

Why do you have the back key on your physical keyboard? As I said before, the home and back keys are the keys of my mobile phone.

**Comment by yyy33** [comment]
（

> > ![Screenshot_20210127-173156589](https://user-images.githubusercontent.com/16245954/106029998-989a3980-60c5-11eb-8dac-5b1228121520.jpg)
> > Make these shortcuts by pressing meta-enter to create a "home" trigger and meta-backspace to create a "back" trigger. I could open Narrator on Windows with the Ctrl+meta+enter shortcut after doing this.
> 
> Why do you have the back key on your physical keyboard? As I said before, the home and back keys are the keys of my mobile phone.

I can't find the back key on my Bluetooth keyboard

> ![Screenshot_20210127-173156589](https://user-images.githubusercontent.com/16245954/106029998-989a3980-60c5-11eb-8dac-5b1228121520.jpg)
> Make these shortcuts by pressing meta-enter to create a "home" trigger and meta-backspace to create a "back" trigger. I could open Narrator on Windows with the Ctrl+meta+enter shortcut after doing this.

I can't find the back key on my Bluetooth keyboard

**Comment by sds100** [comment]
When you press meta-backspace Android treats that shortcut as the "back button". meta-enter is the "home button".

**Comment by sds100** [comment]
So if you press those shortcuts while making a trigger Key Mapper thinks you have pressed the back/home button.

**Comment by yyy33** [comment]
> So if you press those shortcuts while making a trigger Key Mapper thinks you have pressed the back/home button.

Thank you so much. Now I can play computer happily in bed. ha-ha.

**Comment by yyy33** [comment]
> So if you press those shortcuts while making a trigger Key Mapper thinks you have pressed the back/home button.
Will you add this feature to the release version?

**Comment by sds100** [comment]
I will add this to the 2.2 release and release Beta 2 today to the play store
